File: LeetCode/medium/Furthest_Building_You_Can_Reach.cxx
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int b = bricks;
        int l = ladders;
        std::priority_queue<int> q;
        int i = 0;
        for (i = 0; i < heights.size()-1; ++i) {
            
            int diff = heights[i+1]-heights[i];
            if (diff <= 0) 
                continue;
            
            q.push(diff);
            
            b -= diff;
            if (b < 0) {               
                b += q.top();
                q.pop();
                --l;
            }
            if (l < 0)
                break;
        }
        return i;
    }
};

File: LeetCode/medium/Furthest_Building_You_Can_Reach.py
from heapq import heappush, heappop

class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        res = 0
        i = 1
        heap = []
        other = 0
        while i < len(heights):
            d = heights[i] - heights[i-1]
            if d <= 0:
                i += 1
            elif len(heap) < ladders:
                heappush(heap, d)
                i += 1
            else:
                if len(heap) > 00 and d > heap[0]:
                    v = heappop(heap)
                    heappush(heap, d)
                    other += v
                else:
                    other += d
                if other > bricks:
                    break
                else:
                    i += 1
        return i-1
